Wall behavior changing

aleks-papushin opened the pull request.

aleks-papushin committed:
Renamed parameter in difficulty assets, changed and renamed GameWaveDto wall danger related field

aleks-papushin committed:
Wall can't be dangerous more than 1 time in a row

aleks-papushin committed:
Updated .csproj

aleks-papushin committed:
Merged develop and fixed conflict

aleks-papushin committed:
Forgot to add file after conflict fixing

aleks-papushin committed:
Refactored GameManager

aleks-papushin committed:
Implemented arbitrary simultaneous danger wall amount

aleks-papushin committed:
Merge branch 'develop' of https://github.com/aleks-papushin/lamp-bearer into develop

aleks-papushin committed:
Merge branch 'develop' into wall-behavior-changing

aleks-papushin committed:
Tuned waves to introduce new changes with walls and enemy amount

MaXal commented on Assets/Scripts/Game/GameManager.cs:
```diff
@@ -11,22 +11,24 @@ namespace Game
         private GameWaveManager _waveManager;
         private Score _score;
-        private static List<GameObject> WallsToBeDangerous => FindObjectsOfType<WallDanger>().Select(w => w.gameObject).ToList();
+        private int _dangerWallsCurrently;
```
This should be a local variable.

aleks-papushin replied:
It is not static at the last commit and it is used in different coroutines now

aleks-papushin replied:
Agree

MaXal commented on Assets/Scripts/Game/GameManager.cs:
```diff
@@ -59,23 +61,43 @@ private IEnumerator HandleWallsDangerousness()
             {
                 yield return null;
```
Do we really need to skip the first frame?

aleks-papushin replied:
No, it is probably redundant

MaXal commented on Assets/Scripts/Game/GameManager.cs:
```diff
@@ -59,23 +61,43 @@ private IEnumerator HandleWallsDangerousness()
-                StartCoroutine(wall.GetComponent<WallDanger>().BecameDangerousCoroutine(_wallWarningInterval));
+                var wall = _wallsToBeDangerous[new Random().Next(_wallsToBeDangerous.Count)];
+                _dangerWallsCurrently++;
+                wall.CanBeDangerous = false;
```
This should be done inside `BecameDangerousCoroutine` otherwise the details of implementation are exposed.

aleks-papushin replied:
Agree

MaXal commented on Assets/Scripts/Game/GameManager.cs:
```diff
@@ -11,22 +11,24 @@ namespace Game
 {
     public class GameManager : MonoBehaviour
```
Should we call it WallDangerManager since it's the major responsibility of the script? And create a separate instance for OilSpawner?

aleks-papushin replied:
Ok, make sense

MaXal commented on Assets/Scripts/Game/GameManager.cs:
```diff
@@ -59,23 +61,43 @@ private IEnumerator HandleWallsDangerousness()
+                    _waveManager.CurrentWave.dangerWallAmount <= _dangerWallsCurrently ||
+                    _wallsToBeDangerous.Count == 0)
                 {
+                    MoveRestingWallsToDangerous();
```
Looks like two situations are mixed here:
1. We don't need to make a dangerous wall (1 and 2 condition) => we just wait (don't need to call `MoveRestingWallsToDangerous`)
2. We have to make a wall dangerous but we just don't have free one => wait and continue the initial cycle. 

In both cases, the cycle is not needed, simple continue will do fine.

aleks-papushin replied:
Is it outdated comment or not?

aleks-papushin committed:
MoveRestingWallsToDangerous became Coroutine

MaXal commented on Assets/Scripts/Game/GameManager.cs:
```diff
@@ -59,23 +61,43 @@ private IEnumerator HandleWallsDangerousness()
 
-                yield return new WaitForSeconds(_wallCoolDownInterval);
-            }
+        private void MoveWallToRestingWalls(WallDanger wall)
```
Please inline, actions are simple and self-explanatory.

aleks-papushin replied:
Ok

aleks-papushin committed:
Fixed review comments

aleks-papushin committed:
Fixed comment
- extracted OilSpawnManager from GameManager and united with SpawnOil class.
- GameManager renamed to WallDanagerManager

aleks-papushin committed:
Fixed comment after call discussion =)

aleks-papushin committed:
Merge branch 'develop' of https://github.com/aleks-papushin/lamp-bearer into develop

aleks-papushin committed:
Merge branch 'develop' into wall-behavior-changing

aleks-papushin committed:
.csproj files added to .gitignore
